### PR TITLE
Feat/multicallv3

### PR DESCRIPTION
**Pull request by j-vp**

### CODE REVIEW

File: test/fuzz/actions/MultiCall/_MultiCall.fuzz.t.sol [hunk 1]
```diff
@@ -4,12 +4,10 @@
...
 pragma solidity 0.8.19;
 
-import { Fuzz_Test, Constants } from "../../Fuzz.t.sol";
+import { Fuzz_Test } from "../../Fuzz.t.sol";
 
-import { AccountExtension } from "../../../utils/Extensions.sol";
-import { ActionMultiCallV2 } from "../../../../src/actions/MultiCallV2.sol";
-import { ERC20Mock } from "../../../utils/mocks/ERC20Mock.sol";
 import { MultiActionMock } from "../../../utils/mocks/MultiActionMock.sol";
+import { MultiCallExtention } from "../../../utils/extensions.sol";
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/_MultiCall.fuzz.t.sol:10]:
```suggestion
import { MultiCallExtention } from "../../../utils/Extensions.sol";
```

File: src/actions/MultiCall.sol [hunk 2]
```diff
@@ -16,7 +17,10 @@ import { IPermit2 } from "../interfaces/IPermit2.sol";
  * @dev Only calls are used, no delegatecalls.
  * @dev This address will approve random addresses. Do not store any funds on this address!
  */
-contract ActionMultiCall is ActionBase {
+contract ActionMultiCall is ActionBase, ERC721TokenReceiver {
+    address[] internal _mintedAssets;
```
Comment (Thomas-Smets) [inline src/actions/MultiCall.sol:21]: We should be consistent with leading _ for variables.

File: src/actions/MultiCall.sol [hunk 3]
```diff
@@ -55,7 +59,13 @@ contract ActionMultiCall is ActionBase {
             if (depositData.assetTypes[i] == 0) {
                 depositData.assetAmounts[i] = IERC20(depositData.assets[i]).balanceOf(address(this));
             } else if (depositData.assetTypes[i] == 1) {
-                depositData.assetAmounts[i] = 1;
+                if (depositData.assetAmounts[i] == 0) {
```
Comment (Thomas-Smets) [inline src/actions/MultiCall.sol:62]: Maybe add some inline comments whats happening here?

File: src/actions/MultiCall.sol [hunk 4]
```diff
@@ -89,4 +102,29 @@ contract ActionMultiCall is ActionBase {
...
         require(success, string(data));
     }
+
+    /**
+     * @notice Checks the current balance of an asset and ensures it's larger than a required amount.
+     * @param asset The token contract address of the asset that is being checked.
+     * @param minAmountOut The amount of tokens this contract needs to hold at least to succeed.
+     * @dev Can be called as one of the calls in executeAction.
+     */
+    function checkAmountOut(address asset, uint256 minAmountOut) external view {
```
Comment (Thomas-Smets) [inline src/actions/MultiCall.sol:112]: If we don't have something similar for ERC1155, do we need lines 69 to 72?

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+        deployUniswapV3PricingModule(address(univ3PosMgr));
+
+        token0 = new ERC20Mock('Token 0', 'TOK0', 18);
+        token1 = new ERC20Mock('Token 1', 'TOK1', 18);
+        (token0, token1) = token0 < token1 ? (token0, token1) : (token1, token0);
+        token0.mint(users.creatorAddress, 10 ** 38);
+        token1.mint(users.creatorAddress, 10 ** 38);
+
+        // vm.prank(users.creatorAddress);
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:42]: Remove comments

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ */
+pragma solidity 0.8.19;
+
+import { MultiCall_Fuzz_Test } from "./_MultiCall.fuzz.t.sol";
+
+import { ERC20Mock } from "../../.././utils/mocks/ERC20Mock.sol";
+import { NonfungiblePositionManagerMock } from "../../.././utils/mocks/NonfungiblePositionManager.sol";
+
+/**
+ * @notice Fuzz tests for the "checkAmountOut" of contract "MultiCall".
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:13]:
```suggestion
 * @notice Fuzz tests for the "mintUniV3LP" of contract "MultiCall".
```

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+        // vm.prank(users.creatorAddress);
+        // uniV3PricingModule.addAsset(address(univ3PosMgr));
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              TESTS
+    //////////////////////////////////////////////////////////////*/
+
+    function testFuzz_Revert_NotV3(address notV3Contract, bytes4 randomSelector, bytes memory randomBytes) public {
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:50]:
```suggestion
    function testFuzz_Revert_mintUniV3LP_NotV3(address notV3Contract, bytes4 randomSelector, bytes memory randomBytes) public {
```

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        vm.expectRevert(bytes(""));
+        action.mintUniV3LP(notV3Contract, abi.encodeWithSelector(randomSelector, randomBytes));
+    }
+
+    function testFuzz_Success_MintLP(int24 tickLower, int24 tickUpper, uint96 amount0Desired, uint96 amount1Desired)
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:55]:
```suggestion
    function testFuzz_Success_mintUniV3LP_MintLP(int24 tickLower, int24 tickUpper, uint96 amount0Desired, uint96 amount1Desired)
```

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        // vm.prank(users.creatorAddress);
+        // uniV3PricingModule.addAsset(address(univ3PosMgr));
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              TESTS
+    //////////////////////////////////////////////////////////////*/
+
+    function testFuzz_Revert_NotV3(address notV3Contract, bytes4 randomSelector, bytes memory randomBytes) public {
+        vm.expectRevert(bytes(""));
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:51]:
```suggestion
        vm.assume(notV3contract != address(uniV3PosMgr));
        vm.expectRevert(bytes(""));
```

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        token1.mint(users.creatorAddress, 10 ** 38);
+
+        // vm.prank(users.creatorAddress);
+        // uniV3PricingModule.addAsset(address(univ3PosMgr));
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              TESTS
+    //////////////////////////////////////////////////////////////*/
+
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:49]: What happens when you have the correct contract but random selector or bytes?

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    function testFuzz_Success_MintLP(int24 tickLower, int24 tickUpper, uint96 amount0Desired, uint96 amount1Desired)
+        public
+    {
+        token0.mint(address(this), amount0Desired);
+        token1.mint(address(this), amount1Desired);
+        token0.approve(address(univ3PosMgr), amount0Desired);
+        token1.approve(address(univ3PosMgr), amount1Desired);
+
+        amount0Desired = 10 ** 10;
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:63]: What is this sir?

File: test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        amount1Desired = 10 ** 10;
+
+        vm.assume(tickLower < tickUpper);
```
Comment (Thomas-Smets) [inline test/fuzz/actions/MultiCall/MintUniV3LP.fuzz.t.sol:66]: use bound not assume

File: src/actions/MultiCall.sol (hunk 3 above)
Comment (Thomas-Smets) [reply src/actions/MultiCall.sol:62]: + where are the tests for this?

Comment (j-vp) [reply src/actions/MultiCall.sol:62]: test added

Comment (j-vp) [reply src/actions/MultiCall.sol:63]: comments added

**Review by Thomas-Smets — APPROVED** [review]
gtg

**Pull request merged**